### PR TITLE
Add a search capability

### LINKED ISSUE #43: 💡 [Feature]: Add a search capability in the demo listing page
Add a search mechanism in the demo list page to search the demo by its name. Please feel free to open a discussion about your design proposal.

image

### PR DESCRIPTION
**Pull request by omriattiya**
# Description

Add a search mechanism in the demo list page to search the demo by its name

Fixes #43  (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Enter a various of partial texts that contains in the APIs names (such as `api`, `re`, `ion`) and see that the demos list updates accordingly.
- [x] Enter text that doesn't contains in the APIs names (such as `omri`) and see that no demo is presented.
- [x] Delete text and see that all demos are presented.
- [x] Click the clear button and see that all demos are presented.

# Checklist:

- [x] My code follows the [style guidelines](https://github.com/atapas/webapis-playground/blob/master/HOW-TO-ADD-DEMO.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: components/core/SearchBox.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import React, { useEffect, useState } from 'react';
+
+const PLACEHOLDER = 'API Search';
```
Comment (atapas) [inline components/core/SearchBox.js:3]: The text could be, "Search by API names"

File: components/core/App.js [hunk 2]
```diff
@@ -1,27 +1,35 @@
...
-  }, []);
+  };
 
   return (
     <>
+      <SearchBox onSearch={searchAPI} />
       <div className="demo-list">
         {(!loading && demos.length) > 0 &&
-          demos.map((demo, index) => <Demo demo={demo} key={index} />)}
+        demos.map((demo, index) => <Demo demo={demo} key={index} />)}
```
Comment (atapas) [inline components/core/App.js:32]: @omriattiya Can we add a "No matched demo found" warning at the center of the page if it doesn't match criteria? Please feel free to make it as a separate component.

File: components/core/SearchBox.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+export const SearchBox = ({ onSearch }) => {
+
+  const [searchText, setSearchText] = useState('');
```
Comment (atapas) [inline components/core/SearchBox.js:7]: The searchbox seems to be small in width. can we style it to be more larger but responsive?

**Review by atapas — CHANGES_REQUESTED** [review]
@omriattiya Great job!!! I have tested it on the deployed URL.. It works great. I have a few comments to make it really awesome. Would you please consider the requested changes?

**Comment by omriattiya** [comment]
> @omriattiya Great job!!! I have tested it on the deployed URL.. It works great. I have a few comments to make it really awesome. Would you please consider the requested changes?

all resolved :)

File: styles/globals.css [hunk 4]
```diff
@@ -103,22 +103,27 @@ header .links {
...
   font-size: 18px;
-  padding: 10px;
-  margin: 5px;
+  padding: 20px;
+  text-align: center;
+  margin: 20px auto;
   background: rgb(234, 239, 241);
   border: 1px solid rgb(47, 47, 47);
   border-radius: 3px;
+  width: 30vw !important;
```
Comment (atapas) [inline styles/globals.css:126]: Why do we need !important?

Comment (atapas) [reply styles/globals.css:126]: @omriattiya One thong u may want to do is, `width: fit-content;` and make the textbox width 80%

Comment (omriattiya) [reply styles/globals.css]: I removed the `important`.

Comment (omriattiya) [reply styles/globals.css]: > @omriattiya One thing u may want to do is, remove `width: fit-content;` and make the textbox width 80%

that will destroy the UI and move the clear button far away. It's better to make it fit content so the clear button is always relative to it. If you want to change the width of the whole searchbox you need to do it in `searchbox` not in `search-box`

Comment (atapas) [reply styles/globals.css]: @omriattiya I get your point. You are right in handling the cross. I am a bit concerned about the search box is very small on phone. Any idea on fixing it?

image

Comment (omriattiya) [reply styles/globals.css]: added a `min-width` . That should fix it.

**Comment by omriattiya** [comment]
> Hi @omriattiya could you please check why the build fails?

looks like everything is good

**Comment by atapas** [comment]
> > Hi @omriattiya could you please check why the build fails?
> 
> @atapas looks like everything is good

Yes, I redeployed

**Comment by omriattiya** [comment]
@atapas  lets marge?

**Review by atapas — APPROVED** [review]
@omriattiya Changes are looking good. We are good to merge. Thanks for taking care of all the review comments.

**Comment by atapas** [comment]
@omriattiya 

Thanks a lot for taking the issue and making the app better. Please feel free to pick up more, contribute more. I wish you a great #Hactoberfest ahead. You will be added to the contribution list right after we merge this PR. Don't forget to share about your contribution publicly and provide your support with a star ⭐.

Great work 🔥 !!!!

**Comment by atapas** [comment]
@all-contributors please add @omriattiya  for Code

**Pull request merged**